[MRG] fix: avoid overflow in Yeo-Johnson power transform

lsorber opened the pull request:
#### Reference Issues/PRs

Fixes https://github.com/scikit-learn/scikit-learn/issues/23319

#### What does this implement/fix? Explain your changes.

This PR fixes two sources of overflow in the Yeo-Johnson power transform:
1. RuntimeWarning: overflow encountered in multiply from `x_trans_var = x_trans.var()`
2. RuntimeWarning: overflow encountered in power from `out[pos] = (np.power(x[pos] + 1, lmbda) - 1) / lmbda`

The first type of overflow is caused by `np.power`. This PR mitigates this type of overflow by replacing all instances of `np.power` with a numerically more robust formulation based on `np.exp`.

The second type of overflow occurs when the exponents blow up for a marginal gain in log likelihood. This PR mitigates this type of overflow by adding a small regularization term on the exponents.

Non-regression tests for both types of overflow have been added.

lsorber committed:
fix: avoid overflow in Yeo-Johnson power transform

adrinjalali reviewed (commented):
This needs a changelog entry, otherwise LGTM.

lsorber committed:
fix: avoid divide by zero encountered in log

lsorber committed:
docs: add changelog entry

lsorber commented:
Thanks for the review @adrinjalali. I updated this PR with the following changes:

1. I changed `x_trans_exp = np.log(np.abs(x_trans))` to `x_trans_exp = np.log(np.abs(x_trans[x_trans != 0]))` to avoid `RuntimeWarning: divide by zero encountered in log`.
2. I added a changelog entry as requested.
3. I rebased the PR on the latest upstream `main`.

adrinjalali commented:
Hmm, this is changing some values in a docstring we have (hence fail in CI)

adrinjalali commented:
@lsorber please avoid force pushing so that we can see the history of changes. At the end we squash and merge anyway.

lsorber commented:
@adrinjalali Apologies, I thought I had rebased on the latest upstream `main`, but found that wasn't the case and so I did another rebase and force push. I'll avoid doing that from now.

It seems that the doctests fail because they expect to see specific lambda values for the given example. However, because of the new regularisation term on the exponent of the transformed variables, the lamdbas are slightly different from the expected lambdas (see below).

How should we proceed from here?

```
3076 
3077     Examples
3078     --------
3079     &gt;&gt;&gt; import numpy as np
3080     &gt;&gt;&gt; from sklearn.preprocessing import PowerTransformer
3081     &gt;&gt;&gt; pt = PowerTransformer()
3082     &gt;&gt;&gt; data = [[1, 2], [3, 2], [4, 5]]
3083     &gt;&gt;&gt; print(pt.fit(data))
3084     PowerTransformer()
3085     &gt;&gt;&gt; print(pt.lambdas_)
Expected:
    [ 1.386... -3.100...]
Got:
    [ 1.3761854 -3.091368 ]

/home/vsts/work/1/s/sklearn/preprocessing/_data.py:3085: DocTestFailure
```

adrinjalali commented:
maybe @lorentzenchr would have a better idea how to proceed here.

lorentzenchr commented:
I opened #26308 to rely on scipy.stats.yeojohnson.
I think the improvement of this PR should be contributed upstream in scipy. Therefore I close.
